src/transformers/models/vilt/modeling_vilt.py

```diff
@@ -0,0 +1,1268 @@
+
+        valid_nums = [v.size(0) for v in valid_row_idx]
+        non_valid_nums = [v.size(0) for v in non_valid_row_idx]
+        pad_nums = [max_image_length - v for v in valid_nums]
+
+        select = list()
+        for i, (v, nv, p) in enumerate(zip(valid_nums, non_valid_nums, pad_nums)):
+            if p <= 0:
+                valid_choice = torch.multinomial(torch.ones(v).float(), max_image_length)
+                select.append(valid_row_idx[i][valid_choice])
+            else:
+                pad_choice = torch.multinomial(torch.ones(nv).float(), p, replacement=True)
+                select.append(
+                    torch.cat(
+                        [valid_row_idx[i], non_valid_row_idx[i][pad_choice]],
+                        dim=0,
+                    )
+                )
+
+        select = torch.cat(select, dim=0)
+        x = x[select[:, 0], select[:, 1]].view(B, -1, C)
+        x_mask = x_mask[select[:, 0], select[:, 1]].view(B, -1)
+        patch_index = patch_index[select[:, 0], select[:, 1]].view(B, -1, 2)
+        pos_embed = pos_embed[select[:, 0], select[:, 1]].view(B, -1, C)
+
+        cls_tokens = self.cls_token.expand(B, -1, -1)
+        x = torch.cat((cls_tokens, x), dim=1)
+        pos_embed = torch.cat((self.position_embeddings[:, 0, :][:, None, :].expand(B, -1, -1), pos_embed), dim=1)
+        x = x + pos_embed
+        x = self.dropout(x)
+
+        x_mask = torch.cat([torch.ones(x_mask.shape[0], 1).to(x_mask), x_mask], dim=1)
+
+        return x, x_mask, (patch_index, (H, W))
+
+    def forward(
+        self,
+        input_ids,
+        attention_mask,
+        token_type_ids,
+        pixel_values,
+        pixel_mask,
+        inputs_embeds,
+        image_embeds,
+        image_token_type_idx=1,
+    ):
+        # PART 1: text embeddings
+        text_embeds = self.text_embeddings(
+            input_ids=input_ids, token_type_ids=token_type_ids, inputs_embeds=inputs_embeds
+        )
+
+        # PART 2: patch embeddings (with interpolated position encodings)
+        if image_embeds is None:
+            image_embeds, image_masks, patch_index = self.visual_embed(
+                pixel_values, pixel_mask, max_image_length=self.config.max_image_length
+            )
+        else:
+            image_masks = pixel_mask.flatten(1)
+
+        # PART 3: add modality type embeddings
+        # 0 indicates text, 1 indicates image, 2 is optionally used when a second image is provided (NLVR2)
+        if image_token_type_idx is None:
+            image_token_type_idx = 1
+        text_embeds, image_embeds = (
+            text_embeds
+            + self.token_type_embeddings(
+                torch.zeros_like(attention_mask, dtype=torch.long, device=text_embeds.device)
+            ),
+            image_embeds
+            + self.token_type_embeddings(
+                torch.full_like(image_masks, image_token_type_idx, dtype=torch.long, device=text_embeds.device)
+            ),
+        )
```

Comment:
This is too complex. `text_embeds` and `image_embeds` should each incremented on their one lines.